modules/puppet_proxy/configuration_loader.rb

```diff
@@ -0,0 +1,20 @@
+module ::Proxy::Puppet
+  class ConfigurationLoader
+    def load_programmable_settings(settings)
+      raise ::Proxy::Error::ConfigurationError, "Parameter ':puppet_version' is expected to have a non-empty value" if settings[:puppet_version].to_s.empty?
+
+      use_provider = settings[:use_provider]
+      use_provider = [use_provider].compact unless use_provider.is_a?(Array)
+      use_provider << (settings[:puppet_version].to_s >= "4.0" ? :puppet_proxy_puppet_api : :puppet_proxy_legacy)
```

Comment:
Some explicit debug logging to explain which provider was selected and why might be good here, to help point users to the right config file and know which behaviour's in use.